torch/csrc/jit/passes/onnx/preprocess_for_onnx.cpp

```diff
@@ -158,11 +158,89 @@ static void ReplaceAddWithConcat(Block* b) {
   }
 }
 
+// Replace aten::index_put_ with aten::masked_scatter or aten::masked_fill
+// when inputs to the index_put node contains boolean inputs
+//
+// before the pass (index_put -> masked_fill):
+// graph(%0 : Float(2:4, 2:2, 2:1, requires_grad=0, device=cpu)):
+//  %mask.1 : Float(2:4, 2:2, 2:1, requires_grad=0, device=cpu)
+//  %22 : Tensor?[] = prim::ListConstruct(%21)
+//  %23 : Float(requires_grad=0, device=cpu) = prim::Constant[value={1}]()
+//  %24 : bool = prim::Constant[value=0]()
+//  %mask : Float(2:4, 2:2, 2:1) = aten::index_put_(%mask.1, %22, %23, %24)
+//
+// after the pass
+// graph(%0 : Float(2:4, 2:2, 2:1, requires_grad=0, device=cpu)):
+//  %46 : Float(requires_grad=0, device=cpu) = prim::Constant[value={5}]()
+//  %mask.1 : Float(2:4, 2:2, 2:1, requires_grad=0, device=cpu) =
+//  %23 : Float(requires_grad=0, device=cpu) = prim::Constant[value={1}]()
+//  %24 : bool = prim::Constant[value=0]()
+//  %49 : Tensor = aten::masked_fill(%mask.1, %21, %23)
+//
+// before the pass (index_put -> masked_scatter)
+//  %48 : Float(8:1, requires_grad=0, device=cpu) = prim::Constant[value= 1  1
+//                                     1  1  1  1  1  1 [ CPUFloatType{8} ]]()
+//  %42 : Tensor?[] = prim::ListConstruct(%41)
+//  %43 : bool = prim::Constant[value=0]()
+//  %44 : Float(2:4, 2:2, 2:1) = aten::index_put_(%mask, %42, %48, %43)
+//  return (%44)
+//
+// after the pass:
+//  %48 : Float(8:1, requires_grad=0, device=cpu) = prim::Constant[value= 1  1
+//                                     1  1  1  1  1  1 [ CPUFloatType{8} ]]()
+//  %49 : Tensor = aten::masked_fill(%mask.1, %21, %23)
+//  %41 : Bool(2:4, 2:2, 2:1) = aten::to()
+//  %50 : Tensor = aten::masked_scatter(%49, %41, %48)
+//  return (%50)
+static void ReplaceIndexPutWithMaskedScatter(Block* b) {
+  for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end; ++it) {
+    for (auto* child_block : it->blocks()) {
+      ReplaceIndexPutWithMaskedScatter(child_block);
+    }
+    if (it->kind() == aten::index_put_) {
+      auto* lc_node = it->input(1)->node();
+      if (!(lc_node->input(0)->type()->cast<TensorType>()) ||
+          (lc_node->input(0)
+               ->type()
+               ->cast<TensorType>()
+               ->scalarType()
+               .value()) != c10::ScalarType::Bool) {
```

Comment:
This `.value()` needs to be safe guard with `.has_value()`, or checking that `...->scalarType()` is not `c10::nullopt`